s3stream/src/main/java/com/automq/stream/s3/wal/impl/s3/ObjectStorageWALService.java

```diff
@@ -0,0 +1,148 @@
+/*
+ * Copyright 2024, AutoMQ CO.,LTD.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+package com.automq.stream.s3.wal.impl.s3;
+
+import com.automq.stream.s3.ByteBufAlloc;
+import com.automq.stream.s3.operator.ObjectStorage;
+import com.automq.stream.s3.trace.context.TraceContext;
+import com.automq.stream.s3.wal.AppendResult;
+import com.automq.stream.s3.wal.RecoverResult;
+import com.automq.stream.s3.wal.WriteAheadLog;
+import com.automq.stream.s3.wal.common.AppendResultImpl;
+import com.automq.stream.s3.wal.common.RecordHeader;
+import com.automq.stream.s3.wal.common.RecoverResultImpl;
+import com.automq.stream.s3.wal.common.WALMetadata;
+import com.automq.stream.s3.wal.exception.OverCapacityException;
+import com.automq.stream.s3.wal.util.WALUtil;
+import com.automq.stream.utils.Time;
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.CompositeByteBuf;
+import io.netty.buffer.PooledByteBufAllocator;
+import java.io.IOException;
+import java.util.Iterator;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static com.automq.stream.s3.wal.common.RecordHeader.RECORD_HEADER_MAGIC_CODE;
+import static com.automq.stream.s3.wal.common.RecordHeader.RECORD_HEADER_SIZE;
+
+public class ObjectStorageWALService implements WriteAheadLog {
+    private static final Logger log = LoggerFactory.getLogger(ObjectStorageWALService.class);
+
+    protected ObjectStorage objectStorage;
+    protected RecordAccumulator accumulator;
+
+    public ObjectStorageWALService(Time time, ObjectStorage objectStorage, ObjectStorageWALConfig config) {
+        this.objectStorage = objectStorage;
+        this.accumulator = new RecordAccumulator(time, objectStorage, config);
+    }
+
+    @Override
+    public WriteAheadLog start() throws IOException {
+        log.info("Start S3 WAL.");
+        accumulator.start();
+        return this;
+    }
+
+    @Override
+    public void shutdownGracefully() {
+        log.info("Shutdown S3 WAL.");
+        accumulator.close();
+    }
+
+    @Override
+    public WALMetadata metadata() {
+        return null;
```

Comment:
Will we implement this method in the future? If yes, consider throwing an unsupported exception temporarily. If no, return a null may breaking upper logic.